File: vector/src/main/java/im/vector/activity/BugReportActivity.java
/*
 * Copyright 2014 OpenMarket Ltd
 * Copyright 2017 Vector Creations Ltd
 * Copyright 2018 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package im.vector.activity;

import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import org.matrix.androidsdk.util.Log;

import butterknife.BindView;
import im.vector.R;
import im.vector.VectorApp;
import im.vector.util.BugReporter;
import im.vector.util.ThemeUtils;

/**
 * LoggingOutActivity displays an animation while a session log out is in progress.
 */
public class BugReportActivity extends MXCActionBarActivity {
    private static final String LOG_TAG = BugReportActivity.class.getSimpleName();

    @BindView(R.id.bug_report_edit_text)
    EditText mBugReportText;

    @BindView(R.id.bug_report_button_include_logs)
    CheckBox mIncludeLogsButton;

    @BindView(R.id.bug_report_button_include_crash_logs)
    CheckBox mIncludeCrashLogsButton;

    @BindView(R.id.bug_report_button_include_screenshot)
    CheckBox mIncludeScreenShotButton;

    @BindView(R.id.bug_report_progress_view)
    ProgressBar mProgressBar;

    @BindView(R.id.bug_report_progress_text_view)
    TextView mProgressTextView;

    @BindView(R.id.bug_report_scrollview)
    View mScrollView;

    @BindView(R.id.bug_report_mask_view)
    View mMaskView;

    //
    private MenuItem mSendBugReportItem;

    @Override
    public int getLayoutRes() {
        return R.layout.activity_bug_report;
    }

    @Override
    public void initUiAndData() {
        if (null != getSupportActionBar()) {
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        mBugReportText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                refreshSendButton();
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.bug_report, menu);

        CommonActivityUtils.tintMenuIcons(menu, ThemeUtils.getColor(this, R.attr.icon_tint_on_dark_action_bar_color));
        mSendBugReportItem = menu.findItem(R.id.ic_action_send_bug_report);

        refreshSendButton();
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            case R.id.ic_action_send_bug_report:
                sendBugReport();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    protected void onResume() {
        super.onResume();
        refreshSendButton();
    }

    /**
     * Refresh the send button visibility
     */
    private void refreshSendButton() {
        if ((null != mSendBugReportItem) && (null != mBugReportText)) {
            boolean isValid = (null != mBugReportText.getText()) && (mBugReportText.getText().toString().trim().length() > 10);
            mSendBugReportItem.setEnabled(isValid);
            mSendBugReportItem.getIcon().setAlpha(isValid ? 255 : 100);
        }
    }

    /**
     * Send the bug report
     */
    private void sendBugReport() {
        mScrollView.setAlpha(0.3f);
        mMaskView.setVisibility(View.VISIBLE);
        mSendBugReportItem.setEnabled(false);

        mProgressTextView.setVisibility(View.VISIBLE);
        mProgressTextView.setText(getString(R.string.send_bug_report_progress, 0 + ""));

        mProgressBar.setVisibility(View.VISIBLE);
        mProgressBar.setProgress(0);

        BugReporter.sendBugReport(VectorApp.getInstance(),
                mIncludeLogsButton.isChecked(),
                mIncludeCrashLogsButton.isChecked(),
                mIncludeScreenShotButton.isChecked(),
                mBugReportText.getText().toString(),
                new BugReporter.IMXBugReportListener() {
                    @Override
                    public void onUploadFailed(String reason) {
                        try {
                            if (null != VectorApp.getInstance() && !TextUtils.isEmpty(reason)) {
                                Toast.makeText(VectorApp.getInstance(),
                                        VectorApp.getInstance().getString(R.string.send_bug_report_failed, reason), Toast.LENGTH_LONG).show();
                            }
                        } catch (Exception e) {
                            Log.e(LOG_TAG, "## onUploadFailed() : failed to display the toast " + e.getMessage());
                        }

                        mMaskView.setVisibility(View.GONE);
                        mProgressBar.setVisibility(View.GONE);
                        mProgressTextView.setVisibility(View.GONE);
                        mScrollView.setAlpha(1.0f);
                        mSendBugReportItem.setEnabled(true);
                    }

                    @Override
                    public void onUploadCancelled() {
                        onUploadFailed(null);
                    }

                    @Override
                    public void onProgress(int progress) {
                        if (progress > 100) {
                            Log.e(LOG_TAG, "## onProgress() : progress > 100");
                            progress = 100;
                        } else if (progress < 0) {
                            Log.e(LOG_TAG, "## onProgress() : progress < 0");
                            progress = 0;
                        }

                        mProgressBar.setProgress(progress);
                        mProgressTextView.setText(getString(R.string.send_bug_report_progress, progress + ""));
                    }

                    @Override
                    public void onUploadSucceed() {
                        try {
                            if (null != VectorApp.getInstance()) {
                                Toast.makeText(VectorApp.getInstance(), VectorApp.getInstance().getString(R.string.send_bug_report_sent), Toast.LENGTH_LONG)
                                        .show();
                            }
                        } catch (Exception e) {
                            Log.e(LOG_TAG, "## onUploadSucceed() : failed to dismiss the toast " + e.getMessage());
                        }

                        try {
                            BugReportActivity.this.finish();
                        } catch (Exception e) {
                            Log.e(LOG_TAG, "## onUploadSucceed() : failed to dismiss the dialog " + e.getMessage());
                        }
                    }
                });
    }
}

File: vector/src/main/java/im/vector/activity/BugReportActivity.kt
/*
 * Copyright 2018 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package im.vector.activity

import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.core.view.isVisible
import androidx.core.widget.toast
import butterknife.BindView
import butterknife.OnCheckedChanged
import butterknife.OnTextChanged
import im.vector.R
import im.vector.VectorApp
import im.vector.util.BugReporter
import org.matrix.androidsdk.core.Log

/**
 * Form to send a bug report
 */
class BugReportActivity : MXCActionBarActivity() {

    /* ==========================================================================================
     * UI
     * ========================================================================================== */

    @BindView(R.id.bug_report_edit_text)
    lateinit var mBugReportText: EditText

    @BindView(R.id.bug_report_button_include_logs)
    lateinit var mIncludeLogsButton: CheckBox

    @BindView(R.id.bug_report_button_include_crash_logs)
    lateinit var mIncludeCrashLogsButton: CheckBox

    @BindView(R.id.bug_report_button_include_screenshot)
    lateinit var mIncludeScreenShotButton: CheckBox

    @BindView(R.id.bug_report_screenshot_preview)
    lateinit var mScreenShotPreview: ImageView

    @BindView(R.id.bug_report_progress_view)
    lateinit var mProgressBar: ProgressBar

    @BindView(R.id.bug_report_progress_text_view)
    lateinit var mProgressTextView: TextView

    @BindView(R.id.bug_report_scrollview)
    lateinit var mScrollView: View

    @BindView(R.id.bug_report_mask_view)
    lateinit var mMaskView: View

    override fun getLayoutRes() = R.layout.activity_bug_report

    override fun initUiAndData() {
        configureToolbar()

        if (BugReporter.getScreenshot() != null) {
            mScreenShotPreview.setImageBitmap(BugReporter.getScreenshot())
        } else {
            mScreenShotPreview.isVisible = false
            mIncludeScreenShotButton.isChecked = false
            mIncludeScreenShotButton.isEnabled = false
        }
    }

    override fun getMenuRes() = R.menu.bug_report

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        menu.findItem(R.id.ic_action_send_bug_report)?.let {
            val isValid = mBugReportText.text.toString().trim().length > 10
                    && !mMaskView.isVisible

            it.isEnabled = isValid
            it.icon.alpha = if (isValid) 255 else 100
        }

        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.ic_action_send_bug_report -> {
                sendBugReport()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }


    /**
     * Send the bug report
     */
    private fun sendBugReport() {
        mScrollView.alpha = 0.3f
        mMaskView.isVisible = true

        invalidateOptionsMenu()

        mProgressTextView.isVisible = true
        mProgressTextView.text = getString(R.string.send_bug_report_progress, 0.toString() + "")

        mProgressBar.isVisible = true
        mProgressBar.progress = 0

        BugReporter.sendBugReport(VectorApp.getInstance(),
                mIncludeLogsButton.isChecked,
                mIncludeCrashLogsButton.isChecked,
                mIncludeScreenShotButton.isChecked,
                mBugReportText.text.toString(),
                object : BugReporter.IMXBugReportListener {
                    override fun onUploadFailed(reason: String?) {
                        try {
                            if (null != VectorApp.getInstance() && !TextUtils.isEmpty(reason)) {
                                Toast.makeText(VectorApp.getInstance(),
                                        VectorApp.getInstance().getString(R.string.send_bug_report_failed, reason), Toast.LENGTH_LONG).show()
                            }
                        } catch (e: Exception) {
                            Log.e(LOG_TAG, "## onUploadFailed() : failed to display the toast " + e.message, e)
                        }

                        mMaskView.isVisible = false
                        mProgressBar.isVisible = false
                        mProgressTextView.isVisible = false
                        mScrollView.alpha = 1.0f

                        invalidateOptionsMenu()
                    }

                    override fun onUploadCancelled() {
                        onUploadFailed(null)
                    }

                    override fun onProgress(progress: Int) {
                        var progress = progress
                        if (progress > 100) {
                            Log.e(LOG_TAG, "## onProgress() : progress > 100")
                            progress = 100
                        } else if (progress < 0) {
                            Log.e(LOG_TAG, "## onProgress() : progress < 0")
                            progress = 0
                        }

                        mProgressBar.progress = progress
                        mProgressTextView.text = getString(R.string.send_bug_report_progress, progress.toString() + "")
                    }

                    override fun onUploadSucceed() {
                        try {
                            VectorApp.getInstance()?.toast(R.string.send_bug_report_sent, Toast.LENGTH_LONG)
                        } catch (e: Exception) {
                            Log.e(LOG_TAG, "## onUploadSucceed() : failed to dismiss the toast " + e.message, e)
                        }

                        try {
                            finish()
                        } catch (e: Exception) {
                            Log.e(LOG_TAG, "## onUploadSucceed() : failed to dismiss the dialog " + e.message, e)
                        }

                    }
                })
    }

    /* ==========================================================================================
     * UI Event
     * ========================================================================================== */

    @OnTextChanged(R.id.bug_report_edit_text)
    internal fun textChanged() {
        invalidateOptionsMenu()
    }

    @OnCheckedChanged(R.id.bug_report_button_include_screenshot)
    internal fun onSendScreenshotChanged() {
        mScreenShotPreview.isVisible = mIncludeScreenShotButton.isChecked && BugReporter.getScreenshot() != null
    }

    override fun onBackPressed() {
        // Ensure there is no crash status remaining, which will be sent later on by mistake
        BugReporter.deleteCrashFile(this)

        super.onBackPressed()
    }

    /* ==========================================================================================
     * Companion
     * ========================================================================================== */

    companion object {
        private val LOG_TAG = BugReportActivity::class.java.simpleName
    }
}
